File: app/ios/NeuroPulse/Protocol/ProtocolChunker.swift
import Foundation

// BLE protocol blob chunking for Mode 2 protocol upload.
//
// The hub firmware caps a single BLE 5 GATT write at 512 bytes (ATT MTU).
// Compiled NPPS session descriptors can exceed this, so large blobs are split
// into framed chunks. Framing protocol (matches hub firmware reassembly):
//
//   START  (0x01): header byte + 2-byte little-endian total length + ≤509 payload
//   CONT   (0x02): header byte + ≤511 payload
//   END    (0x03): header byte + remaining payload
//   SINGLE (0x04): header byte + payload (whole blob fits in one chunk, ≤509 bytes)
//
// `chunk(_:)` is a pure function (Data → [Data]) with no side effects, so it is
// unit-testable without any BLE stack.
enum ProtocolChunker {

    /// Frame header bytes — must match hub firmware reassembly state machine.
    enum Frame: UInt8 {
        case start  = 0x01
        case cont   = 0x02
        case end    = 0x03
        case single = 0x04
    }

    /// Total BLE 5 ATT MTU write ceiling, in bytes.
    static let maxWriteSize = 512

    /// Max payload in a START chunk: 512 − 1 header − 2 length = 509 bytes.
    /// Also the threshold below/at which a blob ships as a single SINGLE chunk.
    static let maxChunkPayload = 509

    /// Max payload in a CONT or END chunk: 512 − 1 header = 511 bytes.
    private static let maxContPayload = 511

    /// Split a protocol blob into framed BLE-MTU-sized chunks.
    ///
    /// - Empty data → a single empty SINGLE chunk (`[Data([0x04])]`); the hub
    ///   accepts it.
    /// - `data.count ≤ 509` → one SINGLE chunk.
    /// - Otherwise → START chunk (509-byte payload) + zero or more CONT chunks
    ///   (511-byte payload each) + one END chunk with the remainder.
    static func chunk(_ data: Data) -> [Data] {
        // Single-chunk case: fits whole (including the empty blob).
        if data.count <= maxChunkPayload {
            var chunk = Data([Frame.single.rawValue])
            chunk.append(data)
            return [chunk]
        }

        var chunks: [Data] = []

        // START chunk: header + 2-byte LE total length + first 509 payload bytes.
        let total = UInt16(truncatingIfNeeded: data.count)
        var start = Data([Frame.start.rawValue])
        start.append(UInt8(total & 0xFF))          // little-endian low byte
        start.append(UInt8((total >> 8) & 0xFF))   // little-endian high byte
        let startEnd = data.index(data.startIndex, offsetBy: maxChunkPayload)
        start.append(contentsOf: data[data.startIndex..<startEnd])
        chunks.append(start)

        // Remaining payload, split into CONT chunks with an END chunk last.
        var offset = startEnd
        while offset < data.endIndex {
            let remaining = data.distance(from: offset, to: data.endIndex)
            let take = min(maxContPayload, remaining)
            let sliceEnd = data.index(offset, offsetBy: take)
            let isLast = sliceEnd == data.endIndex

            var chunk = Data([(isLast ? Frame.end : Frame.cont).rawValue])
            chunk.append(contentsOf: data[offset..<sliceEnd])
            chunks.append(chunk)

            offset = sliceEnd
        }

        return chunks
    }
}

File: app/android/core/src/main/kotlin/life/neurone/core/session/ProtocolChunker.kt
package life.neurone.core.session

// Port of iOS ProtocolChunker (app/ios/NeurOne/Protocol/ProtocolChunker.swift).
// Splits a compiled + signed session descriptor into BLE-MTU-sized framed chunks for the
// Mode-2 PROTOCOL_UPLOAD write characteristic. Framing (must match hub reassembly):
//   START  (0x01): header + 2-byte little-endian total length + ≤509 payload
//   CONT   (0x02): header + ≤511 payload
//   END    (0x03): header + remaining payload
//   SINGLE (0x04): header + payload (whole blob ≤509 bytes)
// Pure function (ByteArray → List<ByteArray>); no BLE dependency, fully unit-testable.
object ProtocolChunker {

    const val FRAME_START: Byte = 0x01
    const val FRAME_CONT: Byte = 0x02
    const val FRAME_END: Byte = 0x03
    const val FRAME_SINGLE: Byte = 0x04

    /** BLE 5 ATT MTU write ceiling. */
    const val MAX_WRITE_SIZE = 512

    /** Max payload in a START chunk (512 − 1 header − 2 length) and the single-chunk threshold. */
    const val MAX_CHUNK_PAYLOAD = 509

    /** Max payload in a CONT or END chunk (512 − 1 header). */
    private const val MAX_CONT_PAYLOAD = 511

    fun chunk(data: ByteArray): List<ByteArray> {
        // Single-chunk case (fits whole, including the empty blob).
        if (data.size <= MAX_CHUNK_PAYLOAD) {
            return listOf(byteArrayOf(FRAME_SINGLE) + data)
        }

        val chunks = mutableListOf<ByteArray>()

        // START: header + 2-byte LE total length + first 509 payload bytes.
        val total = data.size and 0xFFFF
        val start = byteArrayOf(
            FRAME_START,
            (total and 0xFF).toByte(),          // little-endian low byte
            ((total ushr 8) and 0xFF).toByte(), // little-endian high byte
        ) + data.copyOfRange(0, MAX_CHUNK_PAYLOAD)
        chunks.add(start)

        // Remaining payload → CONT chunks, END chunk last.
        var offset = MAX_CHUNK_PAYLOAD
        while (offset < data.size) {
            val take = minOf(MAX_CONT_PAYLOAD, data.size - offset)
            val sliceEnd = offset + take
            val isLast = sliceEnd == data.size
            val header = if (isLast) FRAME_END else FRAME_CONT
            chunks.add(byteArrayOf(header) + data.copyOfRange(offset, sliceEnd))
            offset = sliceEnd
        }

        return chunks
    }
}
